### PR TITLE
Better command-line parsing

### PR DESCRIPTION
**Pull request by jcburley**
This is a start at re-introducing something akin to `pflag` without (I hope) breaking anything that clients are currently actually using. It refactors how the command line is parsed (see comments on commits for more info), adds some new functionality and error checking, and a fairly decent `--help` output.

It addresses https://github.com/candid82/joker/issues/78 and https://github.com/candid82/joker/issues/76.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -218,75 +218,239 @@ func dialectFromArg(arg string) Dialect {
...
-	workingDir := ""
-	phase := EVAL
-	lint := false
-	dialect := UNKNOWN
-	expr := ""
-	length := len(os.Args) - 1
-	for i := 1; i < length; i++ {
-		switch os.Args[i] {
+func usage() {
+	fmt.Fprintf(os.Stderr, "Joker - %s\n\n", VERSION)
```
Comment (candid82) [inline main.go]: I think help should be printed to stdout, not stderr. Would you agree?

**Comment by jcburley** [comment]
Looks good!

```
[1] craig@doe:~/.go/src/github.com/candid82/joker (better-options-parsing u=)$ ./flag-tests.sh 
[1] craig@doe:~/.go/src/github.com/candid82/joker (better-options-parsing u=)$
```

And I hadn't learned of `gofmt` yet -- it got rid of trailing semicolons i'd removed in later commits (on different branches TBD as PRs).

Will look into the failing test next.

**Comment by jcburley** [comment]
I'm not sure what to think of those failing CircleCI tests -- they all look unrelated to me. Were they passing before??

File: main.go (hunk 1 above)
Comment (jcburley) [reply main.go]: I've seen both approaches taken, but I do prefer `stdout` actually...I've made it easier to change in the future besides changing it now.

**Comment by candid82** [comment]
Linter tests fail for `cljs` files because dialect is not detected correctly. Looks like dialect needs to be initialized to `UNKNOWN` for [auto detection](https://github.com/candid82/joker/pull/83/files#diff-7ddfb3e035b42cd70649cc33393fe32cL290) to kick in.

You can run linter tests locally as well: `./linter-tests.sh`

**Comment by jcburley** [comment]
Sorry to have missed that! I should have been more careful about preserving initialization of the variables (and cross-compared to the proposed `pflag` replacement) instead of assuming `UNKNOWN == 0`.

There are still failures but, if I did things right, I've confirmed they're the same failures as on current `master`.

**Comment by candid82** [comment]
CircleCI seems happy, and I am not seeing any failures in master. What failures are you seeing?

**Comment by candid82** [comment]
I think this is ready to merge. Thanks a lot!

**Comment by jcburley** [comment]
Here's what I was getting:

```
[1] craig@doe:~/.go/src/github.com/candid82/joker (better-options-parsing u+1)$ ./linter-tests.sh 
FAILED: ns-cljs-2 (wrong exit code)
FAILED: ns-cljs-5
EXPECTED:

ACTUAL:
tests/linter/ns-cljs-5/input.cljs:2:30: Exception: Unsupported option(s) supplied: :include-macros

FAILED: threading-macros (wrong exit code)
[1] craig@doe:~/.go/src/github.com/candid82/joker (better-options-parsing * u+1)$
```

I modified the test script to support a `JOKER_EXE` env var to override the Joker executable name, to print the `res` object on `(wrong exit code)`, and also to `(var-set #'exit-code 1)` in that case as well.

Let me know if you want a PR with those changes.

In the meantime, I'll work on merging this latest `master` into my other pending changes so they can be made into clean PRs.

Thanks!

**Comment by jcburley** [comment]
Also fixed https://github.com/candid82/joker/issues/78.

**Comment by candid82** [comment]
Not sure what was causing `linter-tests.sh` to fail. Perhaps you ran it against outdated version of binary? I normally do `./run.sh` first to build the latest version, then `./linter-tests.sh`.

**Pull request merged**